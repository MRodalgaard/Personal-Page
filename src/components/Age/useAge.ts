import { useEffect, useState } from 'react';
import getAge from 'util/getAge';

interface IProps {
  birthday: string;
  updateInterval: number;
}

const DEFAULT_AGE = 0;

const useAge = ({ birthday, updateInterval }: IProps) => {
  const [age, updateAge] = useState(DEFAULT_AGE);

  const actualAge = getAge(birthday) || DEFAULT_AGE;

  useEffect(() => {
    const intervalId = window.setInterval(() => {
      updateAge((ageState) => {
        if (ageState >= actualAge) {
          clearInterval(intervalId);
          return ageState;
        }
        return ageState + 1;
      });
    }, updateInterval);

    return () => {
      clearInterval(intervalId);
    };
  }, [actualAge, updateInterval]);

  return age;
};

export default useAge;
