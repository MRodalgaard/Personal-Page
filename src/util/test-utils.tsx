import { render, RenderOptions } from '@testing-library/react';
import AppProvider from 'components/App/AppProvider';
import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';
import theme from './theme';

type Props = {
  children: ReactNode;
};

const AllTheProviders = ({ children }: Props) => {
  return (
    <AppProvider>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </AppProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, {
    wrapper: AllTheProviders as React.FunctionComponent,
    ...options,
  });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
