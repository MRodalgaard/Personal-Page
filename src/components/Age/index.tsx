import React from 'react';
import styled from 'styled-components';
import useAge from './useAge';

const AgeSpan = styled.span`
  width: 20px;
  display: inline-block;
`;

interface IProps {
  birthday: string;
  updateInterval: number;
}

const Age = (props: IProps) => {
  const age = useAge(props);

  return <AgeSpan>{age}</AgeSpan>;
};

export default Age;
