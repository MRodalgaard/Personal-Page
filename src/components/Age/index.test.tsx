import { mount } from 'enzyme';
import * as React from 'react';
import { act } from 'react-dom/test-utils';
import Age from '.';

describe('Age', () => {
  let tenYearsAgo: string;

  beforeEach(() => {
    const date = new Date();
    date.setFullYear(date.getFullYear() - 10);
    tenYearsAgo = `${date.getMonth()}-${date.getDate()}-${date.getFullYear()}`;
  });

  it('renders', () => {
    const wrapper = mount(<Age birthday={tenYearsAgo} updateInterval={1000} />);
    expect(wrapper.text()).toBe('0');
    expect(wrapper).toMatchSnapshot();
  });

  it('renders final result', () => {
    jest.useFakeTimers();

    const wrapper = mount(<Age birthday={tenYearsAgo} updateInterval={1} />);
    expect(wrapper.text()).toBe('0');

    act(() => {
      jest.runTimersToTime(20);
    });

    expect(wrapper.update().text()).toBe('10');
    wrapper.unmount();
  });
});
