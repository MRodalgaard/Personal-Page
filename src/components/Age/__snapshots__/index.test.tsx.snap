// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Age renders 1`] = `
<Age
  birthday="11-14-2009"
  updateInterval={1000}
>
  <styled.span>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "gEHCiL",
            "rules": Array [
              "
  width: 20px;
  display: inline-block;
",
            ],
          },
          "displayName": "styled.span",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "span",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <span
        className="sc-bdVaJa gEHCiL"
      >
        0
      </span>
    </StyledComponent>
  </styled.span>
</Age>
`;
